ansible/roles/vm_set/library/ptf_control.py

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/python
+
+import json
+import logging
+import traceback
+
+import docker
+
+from ansible.module_utils.debug_utils import config_module_logging
+from ansible.module_utils.basic import AnsibleModule
+
+DOCUMENTATION = '''
+---
+module: ptf_control
+version_added: "0.1"
+author: Xin Wang (xiwang5@microsoft.com)
+short_description: Control PTF container
+description: For controlling PTF container, for example killing processes running in PTF container before stopping it.
+
+Parameters:
+    - ctn_name: Name of the PTF container
+    - command: Command to run, currently only support "kill"
+
+'''
+
+EXAMPLES = '''
+- name: Kill exabgp and ptf_nn_agent processes in PTF container
+  ptf_control:
+    ctn_name: "ptf_vms6-1"
+    command: kill
+'''
+
+
+class PtfControl(object):
+    """This class is for controlling PTF container
+    """
+
+    def __init__(self, module, ctn_name):
+        self.module = module
+        self.ctn_name = ctn_name
+
+        self.pid = PtfControl.get_pid(self.ctn_name)
+
+    def cmd(self, cmdline, use_unsafe_shell=False, ignore_failure=False, verbose=True):
+        rc, out, err = self.module.run_command(cmdline, use_unsafe_shell=use_unsafe_shell)
+        if verbose:
+            msg = {
+                'cmd': cmdline,
+                'rc': rc,
+                'stdout': out.splitlines(),
+                'stderr': err.splitlines()
+            }
+            logging.debug('***** RUN CMD:\n%s' % json.dumps(msg, indent=2))
+
+        if rc != 0 and not ignore_failure:
+            raise Exception("Failed to run command: %s, rc=%d, out=%s, err=%s" % (cmdline, rc, out, err))
+        return rc, out, err
+
+    @staticmethod
+    def get_pid(ctn_name):
+        cli = docker.from_env()
+        try:
+            ctn = cli.containers.get(ctn_name)
+        except Exception:
+            return None
+
+        return ctn.attrs['State']['Pid']
+
+    def get_process_pids(self, process):
+        cmd = 'docker exec -t {} bash -c "pgrep -f \'{}\'"'.format(self.ctn_name, process)
+        _, out, _ = self.cmd(cmd, ignore_failure=True)
+        return [int(pid.strip()) for pid in out.splitlines()]
+
+    def kill_process(self, pid):
+        self.cmd('docker exec -t {} bash -c "kill -9 {}"'.format(self.ctn_name, pid), ignore_failure=True)
+
+    def kill_processes(self):
+        self.cmd('docker exec -t {} bash -c "ps -ef"'.format(self.ctn_name))
```

Comment:
It is for debugging purpose. By default, this module can generate logs like `/tmp/ptf_control_xxx.log`. 
With this command, we can clearly see the processes running in PTF docker before and after the killing. Next time if server crash happens again, we may be able to get some clue from the log.

Recently we noticed some tests may create exabgp processes like "exabgp-psudoswitch1" in PTF container. So, a more aggressive way is required to kill the processes. Also, it would be better to collect more information for debugging if the issue happens again in the future.